File: C++ Library/Union_Find.cpp
#include <iostream>
#include <vector>
using namespace std;

// Union-Find
struct UnionFind {
    vector<int> par, siz;

    // 初期化
    UnionFind(int n) : par(n,-1), siz(n,1) {}

    // 根を求める
    int root(int x) {
        if (par[x] == -1) return x;  // x が根の場合は x を直接返す
        else return par[x] = root(par[x]);  // x の親 par[x] を根に設定する（経路圧縮）
    }

    // x と y が同じグループに属するかどうか（根が一致するかどうか）
    bool issame(int x, int y) {
        return root(x) == root(y);
    }

    // x を含むグループと y を含むグループを併合する
    bool unite(int x, int y) {
        // x, y をそれぞれ根まで移動する
        x = root(x); y = root(y);

        // すでに同じグループのときは何もしない
        if (x==y) return false;

        // union by size （y 側のサイズが小さくなるようにする）
        if (siz[x] < siz[y]) swap(x,y);

        // y を x の子とする
        par[y] = x;
        siz[x] += siz[y];
        return true;
    }

    // x を含むグループのサイズ
    int size(int x) {
        return siz[root(x)];
    }
};

int main() {
    UnionFind uf(7);  // {0},{1},{2},{3},{4},{5},{6}

    uf.unite(1,2);  // {0},{1,2},{3},{4},{5},{6}
    uf.unite(2,3);  // {0},{1,2,3},{4},{5},{6}
    uf.unite(5,6);  // {0},{1,2,3},{4},{5,6}

    cout << uf.issame(1,3) << endl;  // true
    cout << uf.issame(2,5) << endl;  // false

    uf.unite(1,6);  // {0},{1,2,3,5,6},{4}
    cout << uf.issame(2,5) << endl;  // true
}


// --------------------------------------------------------------

// 無向グラフの連結成分の個数を求める
// Union-Find中の「根付き木の根となっている頂点」の個数を数える
int main() {
    // 頂点数と辺数
    int N,M;
    cin >> N >> M;

    // Union-Find を要素数 N で初期化
    UnionFind uf(N);

    // 各辺に対する処理
    for (int i=0; i<M; ++i) {
        int a, b;
        cin >> a >> b;
        uf.unite(a,b);  // a を含むグループと b を含むグループを併合する
    }

    // 集計
    int res = 0;
    for (int x = 0; x < N; ++x) {
        if (uf.root(x) == x) ++res;
    }
    cout << res << endl;
}

File: Python_Library/Union_Find.py
class UnionFind:
    def __init__(self, n):
        self.n = n
        self.par = [i for i in range(n+1)]  # 各要素の親要素の番号を格納するリスト
        self.rank = [0] * (n+1)
        self.size = [1] * (n+1)  # 根（ルート）の場合、そのグループの要素数を格納するリスト

    # 要素 x が属するグループの根を返す
    def find(self, x):
        if self.par[x] == x:
            return x
        else:
            self.par[x] = self.find(self.par[x])
            return self.par[x]

    # 要素 x が属するグループと要素 y が属するグループとを併合する
    def union(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return
        if self.rank[x] < self.rank[y]:
            self.par[x] = y
            self.size[y] += self.size[x]
        else:
            self.par[y] = x
            self.size[x] += self.size[y]
            if self.rank[x] == self.rank[y]:
                self.rank[x] += 1

    # 要素 x, y が同じグループに属するかどうかを返す
    def same_check(self, x, y):
        return self.find(x) == self.find(y)
    
    # 要素 x が属するグループに属する要素をリストで返す
    def members(self, x):
        root = self.find(x)
        return [i for i in range(1,self.n+1) if self.find(i) == root]

    # 要素 x が属するグループのサイズ（要素数）を返す
    def get_size(self, x):
        return self.size[self.find(x)]

    # すべての根の要素をリストで返す
    def roots(self):
        return [i for i, x in enumerate(self.par) if x == i and i != 0]
    
    # グループの数を返す
    def group_count(self):
        return len(self.roots())
    
    # {ルート要素: [そのグループに含まれる要素のリスト], ...} の辞書を返す
    def all_group_members(self):
        return {r: self.members(r) for r in self.roots()}
    
n,q = map(int,input().split())
uf = UnionFind(n)
for i in range(q):
    p,a,b = map(int,input().split())
    if p:
        print(1 if uf.same_check(a,b) else 0)
        continue
    else:
        uf.union(a,b)